x/datadeal/keeper/deal.go

```diff
@@ -0,0 +1,142 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+
+	gogotypes "github.com/gogo/protobuf/types"
+	"github.com/medibloc/panacea-core/v2/x/datadeal/types"
+)
+
+func (k Keeper) CreateDeal(ctx sdk.Context, buyerAddress sdk.AccAddress, deal types.Deal) (uint64, error) {
+
+	dealID, err := k.GetNextDealNumberAndIncrement(ctx)
+	if err != nil {
+		return 0, sdkerrors.Wrapf(err, "failed to get next deal num")
+	}
+
+	newDeal := types.NewDeal(dealID, deal)
+
+	coins := sdk.NewCoins(*deal.GetBudget())
+
+	dealAddress, err := sdk.AccAddressFromBech32(newDeal.GetAddress())
+	if err != nil {
+		return 0, err
+	}
+
+	acc := k.accountKeeper.GetAccount(ctx, dealAddress)
+	if acc != nil {
+		return 0, sdkerrors.Wrapf(types.ErrDealAlreadyExist, "deal %d already exist", dealID)
+	}
+
+	acc = k.accountKeeper.NewAccount(ctx, authtypes.NewModuleAccount(
+		authtypes.NewBaseAccountWithAddress(
+			dealAddress,
+		),
+		newDeal.GetAddress()),
```

Comment:
This should be the name of module, which means `deal{dealID}`, not the deal address.

Reply:
Thanks, I convert it to `strconv.FormatUint(newDeal.Id, 10))`. But, I'm not sure if it's better than address. Can you please explain?

Reply:
When you see `authtypes.NewModuleAccount()`, it accepts the name of module as a second argument.
So in order to make valid module account, you may need to pass the "deal{dealID}" (e.g. "deal1").
You can check the validity of module account as the below test code :)

```go
func (suite dealTestSuite) TestModuleAccount() {
	budget := &sdk.Coin{Denom: assets.MicroMedDenom, Amount: sdk.NewInt(10000000)}

	tempDeal := types.Deal{
		DataSchema:   []string{"http://jsonld.com"},
		Budget:       budget,
		MaxNumData:   10000,
		BuyerAddress: suite.acc1.String(),
	}

	newDeal := types.NewDeal(1, tempDeal)
	dealAddress, err := sdk.AccAddressFromBech32(newDeal.GetAddress())
	suite.Require().NoError(err)
	ma := authtypes.NewModuleAccount(authtypes.NewBaseAccountWithAddress(dealAddress), "deal1")
	suite.Require().NoError(ma.Validate())
}
```

Reply:
Ah, Okay. I understand now. I have to use name when make `NewDealAddress`. I will reflect on the next PR.